File: Olympiad/Baltic/official/boi2014_solutions/coprobber/alternative/coprobber-tree.cpp
#include "coprobber.h"

int cop = 0;
int N;
bool (*A)[MAX_N];

int start(int n, bool a[MAX_N][MAX_N]) {
	N = n;
	A = a;
	return 0;
}

int dist(int v, int to, int parent) {
	if (v == to)
		return 0;
	for (int j = 0; j < N; j++)
		if (j != parent && A[v][j]) {
			int d = dist(j, to, v);
			if (d != -1)
				return 1 + d;
		}
	return -1;
}

int nextMove(int robber) {
	int d = dist(cop, robber, -1);
	for (int j = 0; j < N; j++)
		if (A[cop][j] && dist(j, robber, -1) < d)
			return cop = j;
}

File: Olympiad/Baltic/official/boi2014_solutions/coprobber/alternative/coprobber-tree.pas
unit coprobber;

interface
  const MAX_N = 500;

  type Matrix = array [0..MAX_N - 1, 0..MAX_N - 1] of Boolean;

  function Start(N: Integer; var A: Matrix): Integer;
  function NextMove(R: Integer): Integer;

implementation
var
  NN, i, j, d, cop: Integer;
  AA: Matrix;
function Start(N: Integer; var A: Matrix): Integer;
begin
  NN := N;
  for i := 0 to N - 1 do
    for j := 0 to N - 1 do
      AA[i][j] := A[i][j];
  cop := 0;
  Start := 0;
end;

function dist (v, tto, parent: Integer): Integer;
var j, d: Integer;
begin
  if (v = tto) then begin
    dist := 0;
    exit;
  end;
  for j := 0 to NN - 1 do
    if ((j <> parent) and AA[v][j]) then begin
      d := dist (j, tto, v);
      if (d <> -1) then begin
        dist := d + 1;
        exit;
      end;
    end;
  dist := -1;
end;

function NextMove(R: Integer): Integer;
begin
  d := dist(cop, R, -1);
  for j := 0 to NN - 1 do
    if (AA[cop][j] and (dist(j, R, -1) < d)) then begin
      cop := j;
      NextMove := j;
      exit;
    end;
  
end;

end.
